ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs

```diff
@@ -9,7 +18,59 @@ impl JSONParser {
      * Parses JSON from input URI.
      * Returns the underlying raw image URI, raw animation URI, and JSON.
      */
-    pub async fn parse(_uri: String) -> (Option<String>, Option<String>, Option<Value>) {
-        todo!();
+    pub async fn parse(uri: String) -> (Option<String>, Option<String>, Option<Value>) {
+        match Self::parse_json(uri.clone()).await {
+            Ok(out) => out,
+            Err(e) => {
+                error!("Error parsing JSON: {}", e);
+                (None, None, None)
+            },
+        }
+    }
+
+    /**
+     * Parses JSON from input URI.
+     */
+    async fn parse_json(
+        uri: String,
+    ) -> anyhow::Result<(Option<String>, Option<String>, Option<Value>)> {
+        let (mime, size) = get_uri_metadata(uri.clone()).await?;
+        if ImageFormat::from_mime_type(mime).is_some() {
+            error!("JSON parser received image URI, skipping");
+            return Ok((None, None, None));
+        } else if size > MAX_FILE_SIZE_BYTES {
+            error!("JSON parser received large file, skipping");
+            return Ok((None, None, None));
+        }
+
+        let op = || {
+            async {
+                info!("Sending request for token_uri {}", uri);
+
+                let result = reqwest::get(&uri).await.map_err(Error::transient)?;
+                let parsed_json = result.json::<Value>().await.map_err(Error::transient)?;
+
+                let raw_image_uri = parsed_json["image"].as_str().map(|s| s.to_string());
+                let raw_animation_uri =
+                    parsed_json["animation_url"].as_str().map(|s| s.to_string());
+
+                Ok::<_, Error<reqwest::Error>>((
+                    raw_image_uri,
+                    raw_animation_uri,
+                    Some(parsed_json),
+                ))
+            }
+            .boxed()
+        };
+
+        let backoff = ExponentialBackoff {
+            max_elapsed_time: Some(Duration::from_secs(MAX_RETRY_TIME_SECONDS)),
```

Comment:
so after it exceeds 15 sec it'll fail? how many times is that?

Reply:
Added to constants file, changed to 5 sec. Should be between 3-4 retries [documentation here](https://docs.rs/backoff/latest/backoff/)